libobs: Clamp audio NaN to 0.0f

claudeha opened the pull request:
### Description

Rearrange comparison order so that NaN is caught.
Comparisons with NaN always result in false.
NaNs cause problems later in audio encoder.

### Motivation and Context

When a NaN (not a number) audio sample is fed to OBS, it immediately stops both streaming and recording because ffmpeg audio encoder doesn't like it.

See #4885, which this patch should theoretically fix, if compiler obeys IEEE semantics (no `-ffast-math` or so).

### How Has This Been Tested?

I have not tested yet.

### Types of changes

- Bug fix (non-breaking change which fixes an issue)

### Checklist:

- [ ] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [ ] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

jp9000 commented:
The greater than/less than comparison changes seem unnecessary

Nofew commented:
Media Autobuild Suite builds ffmpeg with -ffast-math by default, so there may be precedent for libffmpeg inside of obs to also not conform to IEEE standards. Can we confirm if this is the situation somehow, since that compiler flag would break this PR?

Testing this patch seems especially prudent to me.

notr1ch commented:
Nice catch! I have confirmed we do potentially have NaN in our `clamp_audio_output` inputs with malformed media files (and likely some glitchy audio devices). This PR does fix one test case with an audio file with out of range signals. It's not known at this time whether the FFmpeg errors users experience are mostly occurring with NaN input or due to FFT overshoot, but this PR definitely seems to fix the NaN input case.

This does mean we need to take care when building to make sure `-ffast-math` is disabled, and provide suitable warning for 3rd party maintainers not to do so.

RytoEX commented:
> The greater than/less than comparison changes seem unnecessary

Ah, I think I see what you mean. The first changed line seems to make the latter two unnecessary. @claudeha can you explain why the second two lines would be needed? Can a NaN get past the first line?

I don't know if we want NaN to clamp to 1.0f as previously proposed or 0.0f as done here. @notr1ch ?

notr1ch commented:
NaN will always fail an equality check: NaN is both greater than 1 and less than 1.

> In comparison operations, positive infinity is larger than all values except itself and NaN, and negative infinity is smaller than all values except itself and NaN. NaN is unordered: it is not equal to, greater than, or less than anything, including itself. x == x is false if the value of x is NaN.

Setting NaN to 0 probably makes the most sense, since NaN should ideally never be present in the first place and likely represents some kind of corrupt / invalid data.

claudeha commented:
On 19/05/2022 23:20, Ryan Foster wrote:
>
>     The greater than/less than comparison changes seem unnecessary
>
> Ah, I think I see what you mean. The first changed line seems to make 
> the latter two unnecessary. @claudeha <https://github.com/claudeha> 
> can you explain why the second two lines would be needed? Can a NaN 
> get past the first line?
>
I guess with the first line (explicitly clamp NaN to 0) the changes in 
the other two would be unnecessary (I changed them first, before 
realizing that clamping NaN to 1.0f might not be such a good idea, and 
didn't change them back after adding the first line).

Possibly it's better to use `isnan()`, which might be less performant 
(though I haven't benchmarked) but could maybe be more robust (though I 
haven't checked -ffast-math behaviour) as well as being more 
self-documenting.

Maybe instead of changing +3-2 lines have a +1-0 with this line added 
before the two unchanged clamping lines:

```
val = isnan(val) ? 0.0f : val;
```

Will need `#include <math.h>` if it's not present already, and also 
glibc needs feature test macros to be defined before include which 
complicates things:

 >       isnan():
 >           _ISOC99_SOURCE || _POSIX_C_SOURCE >= 200112L
 >               || _XOPEN_SOURCE
 >               || /* Since glibc 2.19: */ _DEFAULT_SOURCE
 >               || /* Glibc versions <= 2.19: */ _BSD_SOURCE || 
_SVID_SOURCE


A choice between `isnan(val)` (and associated complexity of build system 
macros) and `! (val == val)` (and associated greater risk that 
`-ffast-math` might break it); then should I make a new PR or force push 
adjusted commit or something else?

notr1ch commented:
`-ffast-math` also breaks `isnan` so I think the current comparison is fine. Generally we prefer force pushes to update PRs unless there's a significant implementation difference that causes it to no longer align with the PR description.

notr1ch commented:
I discovered that `-ffast-math` sets a macro in both GCC and Clang, so we can prevent its use, e.g.

```
#ifdef __FAST_MATH__
#error libobs cannot be built with -ffast-math
#endif
```

RytoEX reviewed (approved):
Looks good to me. Confirmed NaN values are caught and clamped to 0 by this change.

notr1ch commented:
Gonna go ahead and merge this; refusing to build with `-ffast-math` is probably a larger discussion.
